pkg/crane/get.go

```diff
@@ -41,5 +42,46 @@ func getManifest(r string, opt ...Option) (*remote.Descriptor, error) {
 	if err != nil {
 		return nil, fmt.Errorf("parsing reference %q: %v", r, err)
 	}
-	return remote.Get(ref, o.remote...)
+	return getRemoteDescriptor(ref, o)
+}
+
+// getRemoteDescriptor returns the remote.Descriptor of a reference. If the
+// referece points to a index and the platform option is set, instead of the
+// index descriptor, the descriptor of the image which matches the platform is
+// returned.
+func getRemoteDescriptor(ref name.Reference, o options) (*remote.Descriptor, error) {
```

Comment:
Rather than doing this logic here, it might be easier to just do something in `pkg/crane` like:

```go
if opt.platform != nil {
  img, err := desc.Image()
  ...
}
```

in copy and manifest --  I think that would be a smaller change.

Reply:
I don't get what you mean. The logic (for loop below) is required to find the matching platform if the descriptor is a index/list.

Reply:
I'd like to avoid duplicating this logic too much -- this happens in `remote`: https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/v1/remote/index.go#L125-L159

Your loop here assumes there's only ever one level of nesting within an image index, which is usually true, but it's possible to have a [nested index](https://github.com/google/go-containerregistry/tree/72597da50ce903895087828b5fc4b918636366f4/pkg/v1/remote#anatomy-of-an-index), which this wouldn't handle.

If you make `crane.WithPlatform` additionally pass the option down through to `remote.WithPlatform` (similar to [this](https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/crane/options.go#L49)), we can reuse that logic from `remote`.

For copy, you could just change [this](https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/crane/copy.go#L48-L52) to:

```go
if o.platform != nil {
                // If platform is explicitly set, don't copy the whole index, just the appropriate image.
  		if err := copyImage(desc, dstRef, o); err != nil {
			return fmt.Errorf("failed to copy image: %v", err)
		}
} else {
                 if err := copyIndex(desc, dstRef, o); err != nil {
			return fmt.Errorf("failed to copy index: %v", err)
		}
}
```

Since `copyImage` call [`desc.Image`](https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/crane/copy.go#L69), it will do the child matching logic for you.

Fixing `Manifest` and `Digest` would require similar changes.